core/arch/arm/pta/system.c

```diff
@@ -40,6 +52,66 @@ static TEE_Result system_rng_reseed(struct tee_ta_session *s __unused,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result system_derive_ta_unique_key(
+			struct tee_ta_session *s,
+			uint32_t param_types,
+			TEE_Param params[TEE_NUM_PARAMS])
+{
+	size_t data_len = sizeof(TEE_UUID);
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *data = NULL;
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+
+	if (exp_pt != param_types ||
+	    params[0].memref.size > TA_DERIVED_EXTRA_DATA_MAX_SIZE ||
+	    params[1].memref.size < TA_DERIVED_KEY_MIN_SIZE ||
+	    params[1].memref.size > TA_DERIVED_KEY_MAX_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	 /* The derived key shall not end up in non-secure memory by mistake */
+	if (!tee_vbuf_is_sec(params[1].memref.buffer, params[1].memref.size)) {
+		res = TEE_ERROR_SECURITY;
+		goto err;
+	}
+
+	/*
+	 * Check for user provided data that could be mixed together with the
+	 * TA UUID.
+	 */
+	if (params[0].memref.size > 0)
+		if (ADD_OVERFLOW(data_len, params[0].memref.size, &data_len)) {
+			res = TEE_ERROR_SECURITY;
+			goto err;
+		}
+
+	data = calloc(data_len, sizeof(*data));
+	if (!data) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+
+	memcpy(data, &s->ctx->uuid, sizeof(TEE_UUID));
+
+	/* Append the user provided data */
+	if (params[0].memref.size > 0)
+		memcpy(data + sizeof(TEE_UUID), params[0].memref.buffer,
+		       params[0].memref.size);
+
+	res = huk_subkey_derive(HUK_SUBKEY_UNIQUE_TA, data, data_len,
+				params[1].memref.buffer,
+				params[1].memref.size);
+err:
```

Comment:
There's no need to goto this label. The only time point where cleanup is needed is after `calloc()` has successed, but then we'll end up here unconditionally any way.

Reply:
But that means that you'll have rely on `huk_subkey_derive()` failing. Even if that is true (data is null), it feels a bit weird to rely on that instead of just jump out of the function as soon as possible when something went wrong. Also, that means that the memcpy functions will try to write to null? I.e., we will get an abort/panic? Otherwise I agree that if `huk_subkey_derive()` is failing, then the buffer will be cleared twice, which is a bit superfluous. I think I still suggest to keep it, agree?

Reply:
That's not what I meant. All `res = <error_code>; goto err` can be replaced by `return <error_code>`;

Reply:
> All `res = <error_code>; goto err` can be replaced by `return <error_code>`

... **if** we **don't** want to explicitly make sure that the buffer has been zero'ed out when returning. But, I'm having the same argument as my comment in the "huk PR" you did ([comment](https://github.com/OP-TEE/optee_os/pull/2970#discussion_r279335827)). I.e., this is about cryptographic keys, so when such a function fails, we should wipe the buffer where the key where supposed to end up. For "BAD_PARAMETERS" I didn't do it, but to be consequent it should probably also be done from there.

Reply:
Aha, sorry missed that point. Agree, it should be done for "BAD_PARAMETERS" too.

Reply:
If we really want to reveal data only upon successful completion, we should use a temporary buffer and copy back data once OK.

Reply:
But then you'd still return something random back to the user. What we want to return is a known value (like a zero-filled buffer). Having a temporary buffer doesn't seem to help there?

Reply:
I meant output buffer should be left unmodified if invoked command returns with an error code.
If `huk_subkey_derive()` starts to fill the buffer before failing, then buffer late filling with zeros will not ensure previously filled data cannot to read by non-secure world.
As generated key is 16 or 32 byte long, using an allocated temp buffer or even the stack should not be harmful.

Reply:
> then buffer late filling with zeros will not ensure previously filled data cannot to read by non-secure world

But I have check that it's not allowed to use memory from non-secure world. So that shouldn't happen anyways.

> As generated key is 16 or 32 byte long, using an allocated temp buffer or even the stack should not be harmful.

I can for sure do that, but I don't see that is needed. Convince me :)

Reply:
Let's try this: `huk_subkey_derive()` already clears output key buffer when something goes wrong in final step (where data might have been loaded). No need to do that here as output memref was not accessed before `huk_subkey_derive()` is called here.

Reply:
Are you saying that wipe isn't needed at all in this function?

Reply:
Yes, I think so.

Reply:
I've removed all explicit memory clearing even though I personally think it's a bad idea, but since there is a wipe in the `huk_subkey_derive` it's better than nothing.